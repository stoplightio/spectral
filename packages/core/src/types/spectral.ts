import { IDiagnostic, JsonPath } from '@stoplight/types';
import type { JSONSchema7 } from 'json-schema';
import type { Resolver } from '@stoplight/spectral-ref-resolver';

export interface IConstructorOpts {
  resolver?: Resolver;
}

export interface IRunOpts {
  ignoreUnknownFormat?: boolean;
}

export interface ISpectralDiagnostic extends IDiagnostic {
  path: JsonPath;
  code: string | number;
}

export type IRuleResult = ISpectralDiagnostic;

export interface ISpectralFullResult {
  resolved: unknown;
  results: IRuleResult[];
}

export interface IGivenNode {
  path: JsonPath;
  value: unknown;
}

export type JSONSchema = JSONSchema7;
