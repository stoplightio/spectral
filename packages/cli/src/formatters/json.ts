import { Formatter } from './types';

export const json: Formatter = results => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
