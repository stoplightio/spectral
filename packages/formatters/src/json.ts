import { Formatter } from './types';

export const json: Formatter = results => {
  const outputJson = results.map(result => {
    let documentationUrlObject = {};
    if (result.documentationUrl) {
      documentationUrlObject = {
        documentationUrl: result.documentationUrl,
      };
    }
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
      ...documentationUrlObject,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
