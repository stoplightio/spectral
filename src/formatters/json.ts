import { IRuleResult } from '../types';

export const json = (results: IRuleResult[]): string => {
  const outputJson = results.map(result => {
    return {
      code: result.code,
      path: result.path,
      message: result.message,
      severity: result.severity,
      range: result.range,
      source: result.source,
    };
  });
  return JSON.stringify(outputJson, null, '\t');
};
