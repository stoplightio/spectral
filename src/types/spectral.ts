import { IResolveOpts, IResolveResult } from '@stoplight/json-ref-resolver/types';
import { DiagnosticSeverity, Dictionary, IDiagnostic, JsonPath } from '@stoplight/types';
import { JSONSchema4, JSONSchema6, JSONSchema7 } from 'json-schema';
import { IFunction, IProcessedRule, IRule } from '.';
import { Rule } from '../rule';
import { ComputeFingerprintFunc } from '../utils';

export type FunctionCollection = Dictionary<IFunction<any>, string>;
export type RuleCollection = Dictionary<IProcessedRule, string>;
export type PartialRuleCollection = Dictionary<Partial<IRule>, string>;
export type RunRuleCollection = Dictionary<Rule, string>;

export type SpectralDiagnosticSeverity = DiagnosticSeverity | -1;

/**
 * Name of the rule with a boolean value to enable or disable the rule.
 *
 * Will expand on this format later to allow for things like overriding rule options.
 */
export type RuleDeclarationCollection = Dictionary<boolean, string>;

export interface IConstructorOpts {
  resolver?: IResolver;
  computeFingerprint?: ComputeFingerprintFunc;
  useNimma?: boolean;
  proxyUri?: string;
}

export interface IRunOpts {
  ignoreUnknownFormat?: boolean;
  resolve?: {
    documentUri?: string;
  };
}

export interface IRuleResult extends IDiagnostic {
  path: JsonPath;
}

export interface ISpectralFullResult {
  resolved: unknown;
  results: IRuleResult[];
}

export interface IGivenNode {
  path: JsonPath;
  value: any;
}

export type ResolveResult = Omit<IResolveResult, 'runner'>;

export interface IResolver {
  resolve(source: unknown, opts?: IResolveOpts): Promise<ResolveResult>;
}

// todo(@p0lip): make it string | null when working on 6.0
export type FormatLookup = (document: unknown, source?: string) => boolean;
export type RegisteredFormats = Dictionary<FormatLookup, string>;

export type JSONSchema = JSONSchema4 | JSONSchema6 | JSONSchema7;
