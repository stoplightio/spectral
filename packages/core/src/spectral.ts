import { stringify } from '@stoplight/json';
import { DiagnosticSeverity } from '@stoplight/types';
import * as Parsers from '@stoplight/spectral-parsers';
import { createHttpAndFileResolver, Resolver } from '@stoplight/spectral-ref-resolver';
import { memoize } from 'lodash';

import { Document, IDocument, IParsedResult, isParsedResult, ParsedDocument } from './document';
import { DocumentInventory } from './documentInventory';
import { Runner, RunnerRuntime } from './runner';
import { IConstructorOpts, IRunOpts, ISpectralDiagnostic, ISpectralFullResult } from './types';
import { ComputeFingerprintFunc, defaultComputeResultFingerprint } from './utils';
import { Ruleset } from './ruleset/ruleset';
import { generateDocumentWideResult } from './utils/generateDocumentWideResult';
import { ParserOptions, RulesetDefinition } from './ruleset/types';
import { Format } from './ruleset/format';
import { getDiagnosticSeverity } from './ruleset';

memoize.Cache = WeakMap;

export * from './types';

export class Spectral {
  private readonly _resolver: Resolver;

  public ruleset?: Ruleset;

  protected readonly runtime: RunnerRuntime;

  private readonly _computeFingerprint: ComputeFingerprintFunc;

  constructor(protected readonly opts?: IConstructorOpts) {
    this._computeFingerprint = memoize(defaultComputeResultFingerprint);

    if (opts?.resolver !== void 0) {
      this._resolver = opts.resolver;
    } else {
      this._resolver = createHttpAndFileResolver();
    }

    this.runtime = new RunnerRuntime();
  }

  protected parseDocument(target: IParsedResult | IDocument | Record<string, unknown> | string): IDocument {
    return target instanceof Document
      ? target
      : isParsedResult(target)
      ? new ParsedDocument(target)
      : new Document<unknown, Parsers.YamlParserResult<unknown>>(
          typeof target === 'string' ? target : stringify(target, void 0, 2),
          Parsers.Yaml,
        );
  }

  public async runWithResolved(
    target: IParsedResult | IDocument | Record<string, unknown> | string,
    opts: IRunOpts = {},
  ): Promise<ISpectralFullResult> {
    if (this.ruleset === void 0) {
      throw new Error('No ruleset has been defined. Have you called setRuleset()?');
    }

    const document = this.parseDocument(target);
    const ruleset = this.ruleset.fromSource(document.source);

    const inventory = new DocumentInventory(document, this._resolver);
    await inventory.resolve();

    const runner = new Runner(this.runtime, inventory);
    runner.results.push(...this._filterParserErrors(document.diagnostics, ruleset.parserOptions));

    if (document.formats === void 0) {
      const foundFormats = [...ruleset.formats].filter(format => format(inventory.resolved, document.source));
      if (foundFormats.length === 0 && opts.ignoreUnknownFormat !== true) {
        document.formats = null;
        if (ruleset.formats.size > 0) {
          runner.addResult(this._generateUnrecognizedFormatError(document, Array.from(ruleset.formats)));
        }
      } else {
        document.formats = new Set(foundFormats);
      }
    }

    await runner.run(ruleset);
    const results = runner.getResults(this._computeFingerprint);

    return {
      resolved: inventory.resolved,
      results,
    };
  }

  public async run(
    target: IParsedResult | IDocument | Record<string, unknown> | string,
    opts: IRunOpts = {},
  ): Promise<ISpectralDiagnostic[]> {
    return (await this.runWithResolved(target, opts)).results;
  }

  public setRuleset(ruleset: RulesetDefinition | Ruleset): void {
    this.runtime.revoke();
    this.ruleset = ruleset instanceof Ruleset ? ruleset : new Ruleset(ruleset);
  }

  private _generateUnrecognizedFormatError(document: IDocument, formats: Format[]): ISpectralDiagnostic {
    return generateDocumentWideResult(
      document,
      `The provided document does not match any of the registered formats [${formats
        .map(fn => fn.displayName ?? fn.name)
        .join(', ')}]`,
      DiagnosticSeverity.Warning,
      'unrecognized-format',
    );
  }

  private _filterParserErrors(
    diagnostics: ReadonlyArray<ISpectralDiagnostic>,
    parserOptions: ParserOptions,
  ): ISpectralDiagnostic[] {
    return diagnostics.reduce<ISpectralDiagnostic[]>((diagnostics, diagnostic) => {
      if (diagnostic.code !== 'parser') return diagnostics;

      let severity;

      if (diagnostic.message.startsWith('Mapping key must be a string scalar rather than')) {
        severity = getDiagnosticSeverity(parserOptions.incompatibleValues);
      } else if (diagnostic.message.startsWith('Duplicate key')) {
        severity = getDiagnosticSeverity(parserOptions.duplicateKeys);
      } else {
        diagnostics.push(diagnostic);
        return diagnostics;
      }

      if (severity !== -1) {
        diagnostics.push(diagnostic);
        diagnostic.severity = severity;
      }

      return diagnostics;
    }, []);
  }
}
