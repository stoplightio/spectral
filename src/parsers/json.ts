import { getLocationForJsonPath, JsonParserResult, parseWithPointers, trapAccess } from '@stoplight/json';
import { IParser } from './types';

export const parseJson = (input: string) =>
  parseWithPointers(input, { ignoreDuplicateKeys: false, preserveKeyOrder: true });

export const Json: IParser<JsonParserResult<unknown>> = {
  parse: parseJson,
  getLocationForJsonPath,
  trapAccess,
};
