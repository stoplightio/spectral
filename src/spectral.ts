import { safeStringify } from '@stoplight/json';
import { Resolver } from '@stoplight/json-ref-resolver';
import { DiagnosticSeverity, Dictionary } from '@stoplight/types';
import { YamlParserResult } from '@stoplight/yaml';
import { memoize, merge } from 'lodash';

import { STATIC_ASSETS } from './assets';
import { Document, IDocument, IParsedResult, isParsedResult, ParsedDocument } from './document';
import { DocumentInventory } from './documentInventory';
import { functions as defaultFunctions } from './functions';
import * as Parsers from './parsers';
import { readRuleset } from './rulesets';
import { compileExportedFunction } from './rulesets/evaluators';
import { mergeExceptions } from './rulesets/mergers/exceptions';
import { IRulesetReadOptions } from './rulesets/reader';
import { DEFAULT_SEVERITY_LEVEL, getDiagnosticSeverity } from './rulesets/severity';
import { runRules } from './runner';
import {
  FormatLookup,
  FunctionCollection,
  IConstructorOpts,
  IResolver,
  IRuleResult,
  IRunOpts,
  ISpectralFullResult,
  PartialRuleCollection,
  RegisteredFormats,
  RuleCollection,
  RunRuleCollection,
} from './types';
import { IRuleset, RulesetExceptionCollection } from './types/ruleset';
import { ComputeFingerprintFunc, defaultComputeResultFingerprint, empty, prepareResults } from './utils';
import { generateDocumentWideResult } from './utils/generateDocumentWideResult';

memoize.Cache = WeakMap;

export * from './types';

export class Spectral {
  private readonly _resolver: IResolver;

  public functions: FunctionCollection = { ...defaultFunctions };
  public rules: RunRuleCollection = {};
  public exceptions: RulesetExceptionCollection = {};
  public formats: RegisteredFormats;

  private readonly _computeFingerprint: ComputeFingerprintFunc;

  constructor(opts?: IConstructorOpts) {
    this._computeFingerprint = memoize(opts?.computeFingerprint || defaultComputeResultFingerprint);
    this._resolver = opts?.resolver || new Resolver();
    this.formats = {};
  }

  public static registerStaticAssets(assets: Dictionary<string, string>) {
    empty(STATIC_ASSETS);
    Object.assign(STATIC_ASSETS, assets);
  }

  public async runWithResolved(
    target: IParsedResult | IDocument | object | string,
    opts: IRunOpts = {},
  ): Promise<ISpectralFullResult> {
    const document: IDocument =
      target instanceof Document
        ? target
        : isParsedResult(target)
        ? new ParsedDocument(target)
        : new Document<unknown, YamlParserResult<unknown>>(
            typeof target === 'string' ? target : safeStringify(target, undefined, 2),
            Parsers.Yaml,
            opts.resolve?.documentUri,
          );

    if (document.source === void 0 && opts.resolve?.documentUri !== void 0) {
      (document as Omit<Document, 'source'> & { source: string }).source = opts.resolve?.documentUri;
    }

    const inventory = new DocumentInventory(document, this._resolver);
    await inventory.resolve();

    const validationResults: IRuleResult[] = [...inventory.diagnostics, ...document.diagnostics, ...inventory.errors];

    if (document.formats === void 0) {
      const registeredFormats = Object.keys(this.formats);
      const foundFormats = registeredFormats.filter(format => this.formats[format](inventory.resolved));
      if (foundFormats.length === 0 && opts.ignoreUnknownFormat !== true) {
        document.formats = null;
        if (registeredFormats.length > 0) {
          validationResults.push(this._generateUnrecognizedFormatError(document));
        }
      } else {
        document.formats = foundFormats;
      }
    }

    return {
      resolved: inventory.resolved,
      results: prepareResults(
        [...validationResults, ...runRules(inventory, this.rules, this.functions, this.exceptions)],
        this._computeFingerprint,
      ),
    };
  }

  public async run(target: IParsedResult | Document | object | string, opts: IRunOpts = {}): Promise<IRuleResult[]> {
    return (await this.runWithResolved(target, opts)).results;
  }

  public setFunctions(functions: FunctionCollection) {
    empty(this.functions);

    Object.assign(this.functions, { ...defaultFunctions, ...functions });
  }

  public setRules(rules: RuleCollection) {
    empty(this.rules);

    for (const name in rules) {
      if (!rules.hasOwnProperty(name)) continue;
      const rule = rules[name];

      this.rules[name] = {
        name,
        ...rule,
        severity: rule.severity === void 0 ? DEFAULT_SEVERITY_LEVEL : getDiagnosticSeverity(rule.severity),
      };
    }
  }

  public mergeRules(rules: PartialRuleCollection) {
    for (const name in rules) {
      if (!rules.hasOwnProperty(name)) continue;
      const rule = rules[name];
      if (rule) {
        this.rules[name] = merge(this.rules[name], rule);
      }
    }
  }

  private setExceptions(exceptions: RulesetExceptionCollection) {
    const target: RulesetExceptionCollection = {};
    mergeExceptions(target, exceptions);

    empty(this.exceptions);

    Object.assign(this.exceptions, target);
  }

  public async loadRuleset(uris: string[] | string, options?: IRulesetReadOptions) {
    this.setRuleset(await readRuleset(Array.isArray(uris) ? uris : [uris], options));
  }

  public setRuleset(ruleset: IRuleset) {
    this.setRules(ruleset.rules);

    this.setFunctions(
      Object.entries(ruleset.functions).reduce<FunctionCollection>(
        (fns, [key, { code, ref, name, source, schema }]) => {
          if (code === void 0) {
            if (ref !== void 0) {
              ({ code } = ruleset.functions[ref]);
            }
          }

          if (code === void 0) {
            // shall we log or sth?
            return fns;
          }

          fns[key] = compileExportedFunction(code, name, source, schema);
          return fns;
        },
        {
          ...defaultFunctions,
        },
      ),
    );

    this.setExceptions(ruleset.exceptions);
  }

  public registerFormat(format: string, fn: FormatLookup) {
    this.formats[format] = fn;
  }

  private _generateUnrecognizedFormatError(document: IDocument): IRuleResult {
    return generateDocumentWideResult(
      document,
      `The provided document does not match any of the registered formats [${Object.keys(this.formats).join(', ')}]`,
      DiagnosticSeverity.Warning,
      'unrecognized-format',
    );
  }
}
