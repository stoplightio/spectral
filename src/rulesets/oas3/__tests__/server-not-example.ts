import { RuleType, Spectral } from '../../../spectral';
import * as ruleset from '../index.json';

describe('server-not-example.com', () => {
  const s = new Spectral();
  s.addRules({
    'server-not-example.com': Object.assign(ruleset.rules['server-not-example.com'], {
      recommended: true,
      type: RuleType[ruleset.rules['server-not-example.com'].type],
    }),
  });

  test('validate a correct object', async () => {
    const results = await s.run({
      openapi: '3.0.0',
      paths: {},
      servers: [
        {
          url: 'https://stoplight.io',
        },
      ],
    });
    expect(results.length).toEqual(0);
  });

  test('return errors if server is example.com', async () => {
    const results = await s.run({
      openapi: '3.0.0',
      paths: {},
      servers: [
        {
          url: 'https://example.com',
        },
      ],
    });
    expect(results).toEqual([
      expect.objectContaining({
        code: 'server-not-example.com',
        message: 'Server URL should not point at `example.com`.',
        path: ['servers', '0', 'url'],
      }),
    ]);
  });
});
