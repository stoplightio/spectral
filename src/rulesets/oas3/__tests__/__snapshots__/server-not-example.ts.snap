// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server-not-example.com return errors if server is example.com 1`] = `
Array [
  Object {
    "code": "server-not-example.com",
    "message": "must not match the pattern 'example.com'",
    "path": Array [
      "servers",
      0,
      "url",
    ],
    "severity": 1,
    "summary": "Server URL should not point at \`example.com\`.",
  },
]
`;
